File: stripe_terminal/ios/Classes/Mappings/CardMappings.swift
import Foundation
import StripeTerminal

extension CardDetails {
    func toApi() -> CardDetailsApi {
        return CardDetailsApi(
            brand: brand.toApi(),
            country: country,
            expMonth: expMonth,
            expYear: expYear,
            funding: funding.toApi(),
            last4: last4
        )
    }
}

extension CardBrand {
    func toApi() -> CardBrandApi? {
        switch self {
        case .amex:
            return .amex
        case .dinersClub:
            return .dinersClub
        case .discover:
            return .discover
        case .JCB:
            return .jcb
        case .masterCard:
            return .masterCard
        case .unionPay:
            return .unionPay
        case .visa:
            return .visa
        case .unknown:
            return nil
        case .interac:
            return .interac
        case .eftposAu:
            return .eftposAu
        @unknown default:
            fatalError("WTF")
        }
    }
}

extension CardFundingType {
    func toApi() -> CardFundingTypeApi? {
        switch self {
        case .credit:
            return .credit
        case .debit:
            return .debit
        case .prepaid:
            return .prepaid
        case .other:
            return nil
        @unknown default:
            fatalError("WTF")
        }
    }
}


extension CardPresentDetails {
    func toApi() -> CardPresentDetailsApi {
        return CardPresentDetailsApi(
            brand: brand.toApi(),
            cardholderName: cardholderName,
            country: country,
            emvAuthData: emvAuthData,
            expMonth: expMonth,
            expYear: expYear,
            funding: funding.toApi(),
            generatedCard: generatedCard,
            incrementalAuthorizationStatus: incrementalAuthorizationStatus.toApi(),
            last4: last4,
            networks: networks?.toApi(),
            receipt: receipt?.toApi()
        )
    }
}

extension SCPIncrementalAuthorizationStatus {
    func toApi() -> IncrementalAuthorizationStatusApi? {
        switch self {
        case .unknown:
            return nil
        case .notSupported:
            return .notSupported
        case .supported:
            return .supported
        @unknown default:
            fatalError()
        }
    }
}

extension SCPNetworks {
    func toApi() -> CardNetworksApi {
        return CardNetworksApi(
            available: available?.compactMap { 
                guard let brand = CardBrand(rawValue: Int(truncating: $0)) else { return nil }
                return brand.toApi()
            } ?? [],
            preferred: nil
        )
    }
}

extension ReceiptDetails {
    func toApi() -> ReceiptDetailsApi {
        return ReceiptDetailsApi(
            accountType: accountType,
            applicationCryptogram: applicationCryptogram,
            applicationPreferredName: applicationPreferredName,
            authorizationCode: authorizationCode,
            authorizationResponseCode: authorizationResponseCode,
            dedicatedFileName: dedicatedFileName,
            terminalVerificationResults: terminalVerificationResults,
            transactionStatusInformation: transactionStatusInformation
        )
    }
}

// PARAMS

extension CardPresentParametersApi {
    func toHost() throws -> CardPresentParameters {
        let b = CardPresentParametersBuilder()
        if let it = captureMethod { b.setCaptureMethod(it.toHost()) }
        if let it = requestedPriority { b.setRequestedPriority(it.toHost()) }
        if let it = requestExtendedAuthorization { b.setRequestExtendedAuthorization(it) }
        if let it = requestIncrementalAuthorizationSupport { b.setRequestIncrementalAuthorizationSupport(it) }
        return try b.build()
    }
}

extension CardPresentCaptureMethodApi {
    func toHost() -> CardPresentCaptureMethod {
        switch self {
        case .manualPreferred:
            return .manualPreferred
        }
    }
}

extension CardPresentRoutingApi {
    func toHost() -> CardPresentRouting {
        switch self {
        case .domestic:
            return .domestic
        case .international:
            return .international
        }
    }
}

File: stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/CardMappings.kt
package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.models.CardDetails
import com.stripe.stripeterminal.external.models.CardNetworks
import com.stripe.stripeterminal.external.models.CardPresentCaptureMethod
import com.stripe.stripeterminal.external.models.CardPresentDetails
import com.stripe.stripeterminal.external.models.CardPresentParameters
import com.stripe.stripeterminal.external.models.CardPresentRoutingOptionParameters
import com.stripe.stripeterminal.external.models.IncrementalAuthorizationStatus
import com.stripe.stripeterminal.external.models.ReceiptDetails
import com.stripe.stripeterminal.external.models.RoutingPriority
import mek.stripeterminal.api.CardBrandApi
import mek.stripeterminal.api.CardDetailsApi
import mek.stripeterminal.api.CardFundingTypeApi
import mek.stripeterminal.api.CardNetworksApi
import mek.stripeterminal.api.CardPresentCaptureMethodApi
import mek.stripeterminal.api.CardPresentDetailsApi
import mek.stripeterminal.api.CardPresentParametersApi
import mek.stripeterminal.api.CardPresentRoutingApi
import mek.stripeterminal.api.IncrementalAuthorizationStatusApi
import mek.stripeterminal.api.ReceiptDetailsApi

fun CardDetails.toApi(): CardDetailsApi {
    return CardDetailsApi(
        brand = cardBrandToApi(brand),
        country = country,
        expMonth = expMonth.toLong(),
        expYear = expYear.toLong(),
        funding = fundingToApi(funding),
        last4 = last4
    )
}

fun cardBrandToApi(value: String?): CardBrandApi? {
    return when (value) {
        "amex" -> CardBrandApi.AMEX
        "diners" -> CardBrandApi.DINERS_CLUB
        "discover" -> CardBrandApi.DISCOVER
        "jcb" -> CardBrandApi.JCB
        "mastercard" -> CardBrandApi.MASTER_CARD
        "unionpay" -> CardBrandApi.UNION_PAY
        "visa" -> CardBrandApi.VISA
        "unknown" -> null
        else -> null
    }
}

fun fundingToApi(value: String?): CardFundingTypeApi? {
    return when (value) {
        "credit" -> CardFundingTypeApi.CREDIT
        "debit" -> CardFundingTypeApi.DEBIT
        "prepaid" -> CardFundingTypeApi.PREPAID
        "unknown" -> null
        else -> null
    }
}

fun CardPresentDetails.toApi(): CardPresentDetailsApi {
    return CardPresentDetailsApi(
        brand = cardBrandToApi(brand),
        country = country,
        expMonth = expMonth?.toLong() ?: 0L,
        expYear = expYear?.toLong() ?: 0L,
        funding = fundingToApi(funding),
        last4 = last4,
        cardholderName = cardholderName,
        generatedCard = generatedCard,
        receipt = receiptDetails?.toApi(),
        emvAuthData = emvAuthData,
        networks = networks?.toApi(),
        incrementalAuthorizationStatus = incrementalAuthorizationStatus.toApi()
    )
}

fun ReceiptDetails.toApi(): ReceiptDetailsApi {
    return ReceiptDetailsApi(
        accountType = accountType,
        applicationPreferredName = applicationPreferredName,
        authorizationCode = authorizationCode,
        authorizationResponseCode = authorizationResponseCode,
        applicationCryptogram = applicationCryptogram,
        dedicatedFileName = dedicatedFileName,
        transactionStatusInformation = tsi,
        terminalVerificationResults = tvr
    )
}

fun CardNetworks.toApi(): CardNetworksApi {
    return CardNetworksApi(
        available =
        available.mapNotNull { cardBrandToApi(it) },
        preferred = preferred
    )
}

fun IncrementalAuthorizationStatus.toApi(): IncrementalAuthorizationStatusApi? {
    return when (this) {
        IncrementalAuthorizationStatus.NOT_SUPPORTED -> IncrementalAuthorizationStatusApi.NOT_SUPPORTED
        IncrementalAuthorizationStatus.SUPPORTED -> IncrementalAuthorizationStatusApi.SUPPORTED
        IncrementalAuthorizationStatus.UNKNOWN -> null
    }
}

// PARAMS

fun CardPresentParametersApi.toHost(): CardPresentParameters {
    val b = CardPresentParameters.Builder()
    captureMethod?.let { b.setCaptureMethod(it.toHost()) }
    requestExtendedAuthorization?.let { b.setRequestExtendedAuthorization(it) }
    requestIncrementalAuthorizationSupport?.let { b.setRequestIncrementalAuthorizationSupport(it) }
    requestedPriority?.let { b.setRouting(CardPresentRoutingOptionParameters(it.toHost())) }
    return b.build()
}

fun CardPresentCaptureMethodApi.toHost(): CardPresentCaptureMethod {
    return when (this) {
        CardPresentCaptureMethodApi.MANUAL_PREFERRED -> CardPresentCaptureMethod.ManualPreferred
    }
}

fun CardPresentRoutingApi.toHost(): RoutingPriority {
    return when (this) {
        CardPresentRoutingApi.DOMESTIC -> RoutingPriority.DOMESTIC
        CardPresentRoutingApi.INTERNATIONAL -> RoutingPriority.INTERNATIONAL
    }
}
